<a href="https://colab.research.google.com/github/Babisama1/Membuat-design/blob/main/Tes_AI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import torch

nama = "Kelvin"

if torch.cuda.is_available:
  print(f"Turbo aktif, Tuan {nama}")
else:
  print(f"Belum aktif, bro {nama}")

Turbo aktif, Tuan Kelvin


In [ ]:
#mecoba melihat torch bentuknya atau matriks
data_angka = [[1,2,1], [3,4,1], [5,6,1]]

tensor_ai = torch.tensor(data_angka)

print("Data asli = ", data_angka)
print("\nSetelah memakai tensor = \n", tensor_ai)
print("\nBentuknya jika menggunakan tensor = ", tensor_ai.shape)

#ai mencoba menebak angka acak
"""
otak_acak = torch.rand(3,3)
print(f"tebakan acak ai = {otak_acak}")"""

#memindahkan data
device = "cuda" if torch.cuda.is_available() else "cpu"
tensor_turbo = tensor_ai.to(device)
print(f"Sekarang data di : {tensor_turbo.device}")

Data asli =  [[1, 2, 1], [3, 4, 1], [5, 6, 1]]

Setelah memakai tensor = 
 tensor([[1, 2, 1],
        [3, 4, 1],
        [5, 6, 1]])

Bentuknya jika menggunakan tensor =  torch.Size([3, 3])
Sekarang data di : cuda:0


In [ ]:
# MENGAKTIFKAN RADAR
"""#angka 5.0 menjadi titik awal
x = torch.tensor([5.0], requires_grad=True)
#rumus tantangan
y = x**2
#menghitung gradient
y.backward()
#turunan dari x^2 adalah 2x
print(f"Data awalan : {x.item()}")
print(f"Hasil perhitungan : {y.item()}")
print(f"Kompas atau arah AI belajar : {x.grad.item()}")
"""
#THE LOSS atau cara mengetahui tingkatan kesalahan
prediksi_ai = torch.tensor([10.0])
jawaban_asli = torch.tensor([15.0])

error = torch.abs(jawaban_asli - prediksi_ai)

print(f"Ini jawaban dari AI : {prediksi_ai.item()}")
print(f"Ini jawaban asli : {jawaban_asli.item()}")
print(f"Tingkat kesalahan : {error.item()}")

Ini jawaban dari AI : 10.0
Ini jawaban asli : 15.0
Tingkat kesalahan : 5.0


In [ ]:
#AKHIRNYA MEMBUAT AI ASLIIII😁(WALAUPUN MASIH ACAK)
import torch.nn as nn
import torch

lapisan_saraf1= nn.Linear(in_features=1, out_features=1)
#nanti bakal keluar angka random
print(f"Lapisan pengali WEIGHT otomatis : {lapisan_saraf1.weight.item()}")
print(f"Lapisan penambah BIAS  otomatis : {lapisan_saraf1.bias.item()}")

#MELIHAT HASIL AI
input_data = torch.tensor([[5.0]])
hasil_saraf = lapisan_saraf1(input_data)

print(f"Input: {input_data.item()}")
print(f"Hasil setelah lewat saraf : {hasil_saraf.item()}")

#MELIHAT ISI OTAK AI
print("Daftar isi otak AI saat ini : ")
for name, param in lapisan_saraf1.named_parameters():
  print(f"Nama : {name}, nilai : {param.data}")

Lapisan pengali WEIGHT otomatis : -0.4268137216567993
Lapisan penambah BIAS  otomatis : -0.8318283557891846
Input: 5.0
Hasil setelah lewat saraf : -2.9658968448638916
Daftar isi otak AI saat ini : 
Nama : weight, nilai : tensor([[-0.4268]])
Nama : bias, nilai : tensor([-0.8318])


In [ ]:
#PELATIH AI (Stochastic Gradient Descent)
import torch.optim as optim

# 1. Reset sarafnya biar nggak pusing (bikin baru)
lapisan_saraf_fix = nn.Linear(1, 1)
#SGD
pelatih = optim.SGD(lapisan_saraf_fix.parameters(), lr=0.001)
#pengukur salah
pengukur_salah = nn.MSELoss()

#LATIHAN LOOP
input_data = torch.tensor([[1.0]])
target_benar = torch.tensor([[2.0]])

print("--MEMULAI LATIHAN--")

for putaran in range(101):
  #1. AI menebak
  tebakan = lapisan_saraf_fix(input_data)
  #2. berapa salah
  loss = pengukur_salah(tebakan, target_benar)
  #3. bersiihkan catatan lama dan hitung arah perbaikan
  pelatih.zero_grad()
  loss.backward()

  #4. pelatih mulai memperbaiki weight & bias
  pelatih.step()

  #cetak perkembangan setiap 20
  if  putaran % 20 == 0 :
      print(f"Putaran-{putaran} | Tingkat kesalahan : {loss.item():.4f}")

hasil_final = lapisan_saraf_fix(torch.tensor([[1.0]]))
print("Input : 1 | Tebakan harusnya : 2")
print(f"Tebakan AI setelah dilatih : {hasil_final.item():.2f}")

--MEMULAI LATIHAN--
Putaran-0 | Tingkat kesalahan : 7.8152
Putaran-20 | Tingkat kesalahan : 6.6576
Putaran-40 | Tingkat kesalahan : 5.6714
Putaran-60 | Tingkat kesalahan : 4.8313
Putaran-80 | Tingkat kesalahan : 4.1156
Putaran-100 | Tingkat kesalahan : 3.5060
Input : 10 | Tebakan harusnya : 20
Tebakan AI setelah dilatih : 0.14


#PELATIH AI (Stochastic Gradient Descent)
import torch.optim as optim

# 1. Reset sarafnya biar nggak pusing (bikin baru)
lapisan_saraf_fix = nn.Linear(1, 1)
#SGD
pelatih = optim.SGD(lapisan_saraf_fix.parameters(), lr=0.001)
#pengukur salah
pengukur_salah = nn.MSELoss()

…      print(f"Putaran-{putaran} | Tingkat kesalahan : {loss.item():.4f}")

hasil_final = lapisan_saraf_fix(torch.tensor([[1.0]]))
print("Input : 10 | Tebakan harusnya : 20")
print(f"Tebakan AI setelah dilatih : {hasil_final.item():.2f}")

Output:

--MEMULAI LATIHAN--
Putaran-0 | Tingkat kesalahan : 7.8152
Putaran-20 | Tingkat kesalahan : 6.6576
Putaran-40 | Tingkat kesalahan : 5.6714
Putaran-60 | Tingkat kesalahan : 4.8313
Putaran-80 | Tingkat kesalahan : 4.1156
Putaran-100 | Tingkat kesalahan : 3.5060
Input : 10 | Tebakan harusnya : 20
Tebakan AI setelah dilatih : 0.14

📋 Daftar Ceklis Perbaikan:
[✅] Apakah angka Loss sekarang mulai mengecil (misal dari 1.5000 jadi 0.8000 dst)?

[✅] Apakah hasil tebakan akhir sudah mendekati target?